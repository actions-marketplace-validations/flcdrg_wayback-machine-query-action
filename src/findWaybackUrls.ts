import axios from 'axios';

interface IUrlStatusDictionary {
  [index: string]: {
    url: string;
    status: string;
  }[];
}

export interface ILycheeData {
  fail_map: IUrlStatusDictionary;
}

export function parseData(data: string): ILycheeData {
  const parsed: ILycheeData = JSON.parse(data);
  return parsed;
}

export type IReplacements = {
  find: string;
  replace: string;
}[];

export async function findWaybackUrls(
  data: ILycheeData
): Promise<IReplacements> {
  const failedMap = data.fail_map;

  const results: IReplacements = [];

  for (const key in failedMap) {
    if (Object.prototype.hasOwnProperty.call(data.fail_map, key)) {
      const element = data.fail_map[key];

      // look up date in key.
      // TODO Make this configurable
      const regex = /_posts\/(\d+)\/(\d+)-(\d+)-(\d+)-/;
      const matches = regex.exec(key);

      let timestamp: string | undefined;
      if (matches && matches.length === 5) {
        timestamp = matches[2] + matches[3] + matches[4];
      }

      for (const failedItem of element) {
        if (failedItem.status === 'Timeout') {
          const waybackUrl = new URL('https://archive.org/wayback/available');

          waybackUrl.searchParams.append('url', failedItem.url);

          if (timestamp) {
            waybackUrl.searchParams.append('timestamp', timestamp);
          }

          const res = await axios.get(waybackUrl.toString());

          const waybackData: {
            url: string;
            archived_snapshots: {
              closest: {
                status: string;
                available: boolean;
                url: string;
                timestamp: string;
              };
            };
          } = res.data;

          if (waybackData.archived_snapshots) {
            results.push({
              find: waybackData.url,
              replace: waybackData.archived_snapshots.closest.url
            });
          }
        }
      }
    }
  }

  return results;
}
