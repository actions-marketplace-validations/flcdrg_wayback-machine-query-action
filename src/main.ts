import * as core from '@actions/core'
import * as fs from 'fs'
import { findWaybackUrls, parseData } from './findWaybackUrls'

async function run(): Promise<void> {
  try {
    const inputFile: string = core.getInput('input', { required: true});
    const outputFile: string = core.getInput('output');

    const data = readFromFile(inputFile);
    
    if (!data) {
      return
    }

    const parsed = parseData(data);
    const replacements = await findWaybackUrls(parsed)

    const replacementsString = JSON.stringify(replacements);

    core.info(replacementsString);

    fs.writeFile(outputFile, replacementsString, err => {
      core.error(err?.message ?? 'Error writing output file');
    })
    
    core.setOutput('time', new Date().toTimeString())
  } catch (error) {
    if (error instanceof Error) core.setFailed(error.message)
  }
}

run();

function readFromFile(file: string) {
  let r: string | undefined;
  fs.readFile(file, 'utf8' , (err, data) => {
    if (err) {
      core.setFailed(err)
      return;
    }
    r = data;
  });

  return r;
}

