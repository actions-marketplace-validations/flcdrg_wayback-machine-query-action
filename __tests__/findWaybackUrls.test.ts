import { ILycheeData, findWaybackUrls } from '../src/findWaybackUrls';
import { expect, test } from '@jest/globals';

const data: ILycheeData = {
  fail_map: {
    './_posts/2005/2005-03-30-server-and-domain-isolation-using.md': [
      {
        url: 'http://www.microsoft.com/windowsserver2003/technologies/networking/ipsec/default.mspx#EGAA',
        status: 'Timeout'
      }
    ],
    './_posts/2009/2009-03-30-sas-and-typeperf.md': [
      {
        url: 'http://technet.microsoft.com/en-us/library/cc753182.aspx',
        status: 'Failed: Network error'
      },
      {
        url: 'http://www.sqlserver.org.au/',
        status: 'Timeout'
      },
      {
        url: 'http://www.microsoft.com/resources/documentation/windowsnt/4/server/reskit/en-us/reskt4u4/rku4list.mspx?mfr=true',
        status: 'Timeout'
      }
    ]
  }
};

test('finds urls', async () => {
  const result = await findWaybackUrls(data);

  expect(result).toMatchSnapshot();
});

test('finds urls with timestamp', async () => {
  const regex = /_posts\/(\d+)\/(?<year>\d+)-(?<month>\d+)-(?<day>\d+)-/;
  const result = await findWaybackUrls(data, regex);

  expect(result).toMatchSnapshot();
});
