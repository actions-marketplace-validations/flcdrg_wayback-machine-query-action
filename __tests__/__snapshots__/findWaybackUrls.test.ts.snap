// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`finds urls 1`] = `
Array [
  Object {
    "find": "http://www.microsoft.com/windowsserver2003/technologies/networking/ipsec/default.mspx#EGAA",
    "replace": "http://web.archive.org/web/20050404133316/http://www.microsoft.com:80/windowsserver2003/technologies/networking/ipsec/default.mspx",
  },
  Object {
    "find": "http://www.sqlserver.org.au/",
    "replace": "http://web.archive.org/web/20090217192805/http://www.sqlserver.org.au:80/",
  },
  Object {
    "find": "http://www.microsoft.com/resources/documentation/windowsnt/4/server/reskit/en-us/reskt4u4/rku4list.mspx?mfr=true",
    "replace": "http://web.archive.org/web/20091206231445/http://www.microsoft.com:80/resources/documentation/windowsnt/4/server/reskit/en-us/reskt4u4/rku4list.mspx?mfr=true",
  },
]
`;
